pylint/checkers/refactoring/recommendation_checker.py

```diff
@@ -83,21 +83,23 @@ def _check_consider_iterating_dictionary(self, node: nodes.Call) -> None:
             return
         if node.func.attrname != "keys":
             return
-        if (
-            isinstance(node.parent, (nodes.For, nodes.Comprehension))
-            or isinstance(node.parent, nodes.Compare)
-            and any(
-                op
-                for op, comparator in node.parent.ops
-                if op == "in" and comparator is node
-            )
+        inferred = utils.safe_infer(node.func)
+        if not isinstance(inferred, astroid.BoundMethod) or not isinstance(
+            inferred.bound, nodes.Dict
         ):
-            inferred = utils.safe_infer(node.func)
-            if not isinstance(inferred, astroid.BoundMethod) or not isinstance(
-                inferred.bound, nodes.Dict
-            ):
-                return
+            return
 
+        if isinstance(node.parent, (nodes.For, nodes.Comprehension)):
+            self.add_message("consider-iterating-dictionary", node=node)
+
+        while not isinstance(node.parent, (nodes.Compare, nodes.Module)):
+            node = node.parent
```

Comment:
Let's use something like this here:
https://github.com/PyCQA/astroid/blob/907818246c89ef938b23448e0cecc52b7c8693e0/astroid/nodes/node_classes.py#L4830-L4835

I think something like this should work:
```python
T_Node = TypeVar("T_Node", bound=NodeNG)

def get_if_node_type_ancestor(node: NodeNG, type: Tuple[Type[T_Node]]) -> Optional[T_Node]:
    """Return the first parent node that is any of the provided types (or None)"""
    for ancestor in node.node_ancestors():
        if isinstance(ancestor, type):
            return ancestor
    return None
```

Note that `type` is a `tuple` so we can use this function later to also check for multiple types.

We could at it to the checker utils file.

WIth this, we could do one check for ancestors and move all of the logic after that check

Reply:
really smart, i like that